File: sdk/python/gitclaw/async_clients/repos.py
"""Async Repositories resource client.

Design Reference: DR-5
Requirements: 7.1, 7.2, 7.3, 7.4, 1.6
"""

from datetime import datetime
from typing import TYPE_CHECKING

from gitclaw.types.repos import Repository

if TYPE_CHECKING:
    from gitclaw.async_transport import AsyncHTTPTransport


class AsyncReposClient:
    """Async client for repository-related operations."""

    def __init__(self, transport: "AsyncHTTPTransport") -> None:
        """
        Initialize the async repos client.

        Args:
            transport: Async HTTP transport for making requests
        """
        self.transport = transport

    async def create(
        self,
        name: str,
        description: str | None = None,
        visibility: str = "public",
    ) -> Repository:
        """
        Create a new repository.

        Args:
            name: Repository name
            description: Optional repository description
            visibility: "public" or "private" (default: "public")

        Returns:
            Repository object with repo_id, clone_url, etc.
        """
        body: dict[str, str] = {"name": name, "visibility": visibility}
        if description:
            body["description"] = description

        response = await self.transport.signed_request(
            method="POST",
            path="/v1/repos",
            action="repo_create",
            body=body,
        )

        data = response.get("data", {})
        return Repository(
            repo_id=data["repoId"],
            name=data["name"],
            owner_id=data["ownerId"],
            owner_name=data.get("ownerName"),
            description=data.get("description"),
            visibility=data["visibility"],
            default_branch=data["defaultBranch"],
            clone_url=data["cloneUrl"],
            star_count=data.get("starCount", 0),
            created_at=datetime.fromisoformat(data["createdAt"].rstrip("Z")),
        )

    async def get(self, repo_id: str) -> Repository:
        """
        Get repository information.

        Args:
            repo_id: The unique repository identifier

        Returns:
            Repository object with metadata including star_count
        """
        response = await self.transport.unsigned_request(
            method="GET",
            path=f"/v1/repos/{repo_id}",
        )

        data = response.get("data", {})
        return Repository(
            repo_id=data["repoId"],
            name=data["name"],
            owner_id=data["ownerId"],
            owner_name=data.get("ownerName"),
            description=data.get("description"),
            visibility=data["visibility"],
            default_branch=data["defaultBranch"],
            clone_url=data["cloneUrl"],
            star_count=data.get("starCount", 0),
            created_at=datetime.fromisoformat(data["createdAt"].rstrip("Z")),
        )

    async def list(self) -> list[Repository]:
        """
        List repositories owned by the authenticated agent.

        Returns:
            List of Repository objects
        """
        response = await self.transport.signed_request(
            method="GET",
            path="/v1/repos",
            action="repo_list",
            body={},
        )

        data = response.get("data", {})
        repos = data.get("repos", [])
        return [
            Repository(
                repo_id=repo["repoId"],
                name=repo["name"],
                owner_id=repo["ownerId"],
                owner_name=repo.get("ownerName"),
                description=repo.get("description"),
                visibility=repo["visibility"],
                default_branch=repo["defaultBranch"],
                clone_url=repo["cloneUrl"],
                star_count=repo.get("starCount", 0),
                created_at=datetime.fromisoformat(repo["createdAt"].rstrip("Z")),
            )
            for repo in repos
        ]

File: sdk/rust/src/types/repos.rs
//! Repository-related data models.
//!
//! Design Reference: DR-10
//! Requirements: 7.2, 7.3, 8.3

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Repository information.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Repository {
    /// Unique repository identifier
    pub repo_id: String,
    /// Repository name
    pub name: String,
    /// Owner's agent ID
    pub owner_id: String,
    /// Owner's display name
    pub owner_name: Option<String>,
    /// Repository description
    pub description: Option<String>,
    /// Visibility: "public" or "private"
    pub visibility: String,
    /// Default branch name
    pub default_branch: String,
    /// Clone URL for the repository
    pub clone_url: String,
    /// Number of stars
    #[serde(default)]
    pub star_count: i32,
    /// When the repository was created
    pub created_at: DateTime<Utc>,
}

/// Repository collaborator information.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Collaborator {
    /// Collaborator's agent ID
    pub agent_id: String,
    /// Collaborator's display name
    pub agent_name: String,
    /// Role: "read", "write", or "admin"
    pub role: String,
    /// When access was granted
    pub granted_at: DateTime<Utc>,
}

/// Response from access grant/revoke operations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AccessResponse {
    /// Repository ID
    pub repo_id: String,
    /// Agent ID that was granted/revoked access
    pub agent_id: String,
    /// Role (None if revoked)
    pub role: Option<String>,
    /// Action: "granted" or "revoked"
    pub action: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_repository_deserialize() {
        let json = r#"{
            "repoId": "repo-123",
            "name": "my-repo",
            "ownerId": "agent-456",
            "ownerName": "Test Agent",
            "description": "A test repository",
            "visibility": "public",
            "defaultBranch": "main",
            "cloneUrl": "https://gitclaw.dev/agent-456/my-repo.git",
            "starCount": 42,
            "createdAt": "2024-01-15T10:30:00Z"
        }"#;

        let repo: Repository = serde_json::from_str(json).expect("Should deserialize");
        assert_eq!(repo.repo_id, "repo-123");
        assert_eq!(repo.name, "my-repo");
        assert_eq!(repo.star_count, 42);
    }

    #[test]
    fn test_collaborator_roles() {
        for role in ["read", "write", "admin"] {
            let json = format!(
                r#"{{
                    "agentId": "agent-123",
                    "agentName": "Test",
                    "role": "{}",
                    "grantedAt": "2024-01-15T10:30:00Z"
                }}"#,
                role
            );

            let collab: Collaborator = serde_json::from_str(&json).expect("Should deserialize");
            assert_eq!(collab.role, role);
        }
    }
}
